### PR TITLE
fix(auth): stop writing activation token expiry as ISO 'T' — closes #670, #673

### PR DESCRIPTION
**Pull request by BreakableHoodie**
## Summary

Fixes the SEC-F1-class bug where `activation_token_expires_at` was written as a raw ISO string (`T` separator) but compared in SQL against D1's space-separated `datetime('now')`, making the SQL expiry guard in `activate.js` permanently inert (defense-in-depth only — a separate JS check still caught real expiry, so this was not exploitable). Fixing the write side surfaced a second bug in that same JS check — it now parses the new format as local time on non-UTC hosts — which is also fixed here. Also adds first-ever direct test coverage for `turnstile.js`, `crypto.js`, negative cases for `totp.js`, and the `_middleware.js` PRAGMA foreign_keys allowlist.

Closes #670
Closes #673

## What changed

| File | Change |
|------|--------|
| `functions/api/admin/auth/signup.js` | `activation_token_expires_at` now written via `toSqliteDateTime()` instead of raw `.toISOString()` |
| `functions/api/auth/resend-activation.js` | Same fix at its own write site |
| `functions/utils/authAttempts.js` | Added `fromSqliteDateTime()` — the inverse parser, safe for both the new space-separated format and legacy ISO `'T'`/`'Z'` rows |
| `functions/api/auth/activate.js` | JS-level expiry check now uses `fromSqliteDateTime()` instead of a bare `new Date(value)`, which silently fell back to local-time parsing once the stored value stopped carrying a `'Z'` suffix |
| `functions/api/admin/auth/__tests__/signup.test.js` | New regression test: reads the real column back off the test DB after signup's own INSERT, asserts no `T` / correct shape |
| `functions/api/auth/__tests__/resend-activation.test.js` | New file: same regression test for resend-activation's write site, plus baseline coverage (already-active user, unknown email) |
| `functions/api/auth/__tests__/activate.test.js` | New file (activate.js had zero coverage before): valid/expired paths for both the new SQLite-format and legacy-ISO expiry values, unknown token, already-activated account |
| `functions/utils/__tests__/turnstile.test.js` | New file — see coverage list below |
| `functions/utils/__tests__/crypto.test.js` | New file — see coverage list below |
| `functions/utils/__tests__/totp.test.js` | Negative cases added; existing RFC 6238 vector tests untouched |
| `functions/__tests__/middlewarePragma.test.js` | New file — see coverage list below |
| `CLAUDE.md` | Documented PBKDF2's 600,000-iteration default (a test comment cited this value but the doc didn't actually state it) |

### Test coverage added per file (#673)

**`turnstile.test.js`** (14 tests): fails closed in production when `TURNSTILE_SECRET_KEY` is unset (the critical case CLAUDE.md names explicitly); allows the request only when `isDevRequest()` is true, and confirms it never trusts client-supplied headers for that decision; returns `false` without calling `fetch` when the token is missing/non-string; returns `true` only on a 2xx response with `success: true`, `false` on 2xx+`success: false`; returns `false` and never reads the body when siteverify responds non-2xx (#682's early return); returns `false` and logs when `fetch` throws or the body isn't valid JSON; forwards secret/token/client-IP to siteverify correctly.

**`crypto.test.js`** (10 tests): `pbkdf2$iterations$salt$hash` format; documented iteration count (600,000); hash/verify roundtrip including wrong-password rejection; salt uniqueness across two hashes of the same password; five malformed-hash negative cases (wrong segment count, non-numeric iterations, empty salt/digest, unrecognized string, garbage base64) proving `verifyPassword` never throws on corrupt input.

**`totp.test.js`** (+6 new, existing RFC vectors untouched): `verifyTotp` returns `false` for a wrong code, a code outside the time window, and an empty/null code; `verifyBackupCode` returns `false` for an invalid code (without consuming anything) and for an already-consumed code (re-checked against the trimmed `remaining` list, mirroring the real caller pattern).

**`middlewarePragma.test.js`** (8 tests): `PRAGMA foreign_keys = ON` fires for POST/PUT/PATCH/DELETE and an unrecognized method (proving it's an allowlist of read-only methods, not a denylist of known mutators); does NOT fire for GET/HEAD; doesn't throw when `env.DB` is absent.

**`activate.test.js`** (6 tests, net-new file): valid non-expired token (both new SQLite-format and legacy-ISO expiry) activates; expired token (both formats) is rejected; unknown token and already-activated account are rejected.

## Security / correctness notes

- The original bug was **not exploitable**: `activate.js`'s JS-level check (`new Date(...) < new Date()`) still caught real expiry before this PR, so the inert SQL guard was a silent loss of defense-in-depth, not a live bypass. Confirmed by red-then-green testing (see below).
- Fixing the write format surfaced a **second, related bug**: once the stored value stopped carrying a `Z` suffix, the JS check itself became TZ-ambiguous (`new Date("YYYY-MM-DD HH:MM:SS")` falls back to local-time parsing in V8). Verified empirically — `TZ=America/Toronto` shifted a parsed instant 4 hours later than the correct UTC time. Not exploitable in production (Cloudflare's `workerd` runtime always reports UTC), but it's a real correctness bug for any non-UTC host and for test reliability, and is fixed via the new `fromSqliteDateTime()` helper.
- No migration needed: legacy ISO-formatted rows still parse correctly through `fromSqliteDateTime()`'s `includes("T")` branch, and 24h-lived tokens self-clear the mixed-format window.
- No production behavior change to the bot gate (`turnstile.js`), CSRF, or crypto primitives — this PR is test-only for `turnstile.js`/`crypto.js`/`totp.js`/`_middleware.js`.

### Red-then-green verification (issue #670's explicit ask)

Both new regression assertions were verified to fail against the pre-fix code, not just "fail by construction":

1. **Write-site tests** (`signup.test.js`, `resend-activation.test.js`): reverted `signup.js`/`resend-activation.js` via `git stash`, reran — both new assertions failed red (`expected '2026-07-30T00:07:56.425Z' to match /^\d{4}-\d{2}-\d{2} ...$/`). Restored the fix, reran — green.
2. **`activate.js` TZ-parsing fix**: reverted `activate.js` via `git stash` under `TZ=America/Toronto`, reran `activate.test.js` — the "expired token" test failed red, though interestingly via a *different* code path than expected: the JS check under-detected expiry (local-time drift made a 1-hour-past expiry look 3 hours in the future), but the **SQL-level guard** (independently fixed by this same PR's write-side change) still caught it and rejected with a different message (`"Invalid or expired token"` instead of `"Activation link expired"`). This is a good demonstration of defense-in-depth actually working — the two layers compensated for each other — but the JS-level message/behavior was still wrong, which the fix corrects. Restored the fix, reran under the same `TZ` — green.

One deliberate technique deviation from #670's issue text worth flagging: the issue asked to mirror the capturing-DB technique from `functions/utils/__tests__/invariants.test.js`. I used the real better-sqlite3 test DB instead (the pattern already established in `signup.test.js`) and read the persisted column back after the handler's actual INSERT/UPDATE — this exercises the full write path (bind + real SQLite storage) rather than mocking it, which for these particular multi-query handlers was both simpler and at least as strong a proof of the binding path.

### Follow-ups filed (not blocking this PR)

- #686 — six more write sites (`tokens.js`, `trustedDevice.js`, `share.js`, `invite-codes.js`, `users.js`, `login.js`) and three more read sites still inline the `toSqliteDateTime`/`fromSqliteDateTime`-equivalent logic instead of importing the shared helpers.
- #687 — `signup.test.js`'s pre-existing invite-code expiry fixtures use raw `.toISOString()` and the "expired invite code" test currently only passes because of date-prefix divergence, not real time comparison (found during review; not introduced by this PR).

## Verification

- [x] Tests: 871 pass, 0 failures, 6 pre-existing todo (`npm test`)
- [x] ESLint: 0 errors (`npm run lint`)
- [x] Format check: clean (`npm run format:check`)
- [x] Build: N/A — backend-only change, frontend untouched (`git diff --stat -- frontend/` is empty)
- [x] `validate:openapi`: N/A — `docs/api-spec.yaml` not touched
- [x] Manual smoke: N/A — covered by the red/green regression tests above; no browser-facing surface changed

---
Built by Sonny · Reviewed by Theo · 🤖 [Claude Code](https://claude.ai/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved account activation and resend-activation by normalizing token-expiry dates to a SQLite-compatible format and failing closed on malformed/expired values.
  - Hardened password verification to respect embedded PBKDF2 iteration counts and safely reject malformed stored hashes.
  - Ensured foreign-key enforcement is applied to write requests, while reads and preflights remain unaffected.
  - Strengthened Turnstile and TOTP/backup-code negative-case handling.  
- **Documentation**
  - Clarified PBKDF2 stored-hash format and default iteration count, including legacy compatibility.  
- **Tests**
  - Added/expanded Vitest coverage for middleware PRAGMA behaviour, activation/resend expiry guards, crypto/TOTP/Turnstile edge cases.

### CODE REVIEW

File: functions/__tests__/middlewarePragma.test.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    prepare(sql) {
+      return {
+        async run() {
+          executed.push(sql);
+          return { success: true, meta: { changes: 0 } };
+        },
+      };
+    },
+  };
+  return { DB, executed };
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/__tests__/middlewarePragma.test.js:38]: Fixed in bae65e6 — `run()` now awaits `Promise.resolve()` before pushing to `executed`, so the ordering assertion proves the PRAGMA statement *completed* before `next()`, not merely that it was invoked.

File: functions/__tests__/middlewarePragma.test.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    const { DB, executed } = dbWithPragmaSpy();
+    const seen = {};
+    const request = new Request(NEUTRAL_URL, { method: "OPTIONS", headers: LOOPBACK_HEADERS });
+
+    await onRequest({ request, env: { DB }, data: {}, next: okNext(executed, seen) });
+
+    expect(executed).not.toContain(PRAGMA);
+    // Short-circuited: the downstream handler is never invoked.
+    expect(seen.called).toBeUndefined();
+  });
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/__tests__/middlewarePragma.test.js:129]: Fixed in bae65e6 — the OPTIONS test now captures the `onRequest` response and asserts `response.status` is `204`, alongside the existing PRAGMA-absence and downstream-handler assertions.

File: functions/api/auth/activate.js [hunk 3]
```diff
@@ -48,7 +50,7 @@ export async function onRequestPost(context) {
       );
     }
 
-    if (user.activation_token_expires_at && new Date(user.activation_token_expires_at) < new Date()) {
+    if (user.activation_token_expires_at && fromSqliteDateTime(user.activation_token_expires_at) < new Date()) {
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/api/auth/activate.js]: Already addressed in 839af59, prior to this review pass — `fromSqliteDateTime()` parses once and returns `null` for unparseable input, and the caller (`activate.js`) treats `null` as expired instead of proceeding to the SQL update. Covered by the "rejects a malformed activation_token_expires_at instead of failing open" test. Re-verified against current code in this pass — no further change needed.

File: functions/utils/__tests__/totp.test.js [hunk 4]
```diff
@@ -43,4 +43,63 @@ describe("totp utilities", () => {
...
+    });
+
+    it("verifyTotp returns false for a code outside the time window", async () => {
+      const secret = generateTotpSecret();
+      // 10 minutes away from "now" is well outside the default ±1 step
+      // (±30s) window, so this code must not validate against the current
+      // time that verifyTotp checks internally.
+      const farFutureCode = await generateTotpCode(secret, Date.now() + 10 * 60 * 1000);
+
+      expect(await verifyTotp(secret, farFutureCode)).toBe(false);
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/utils/__tests__/totp.test.js]: Fixed in bae65e6 — the test now freezes time via `vi.useFakeTimers()`/`vi.setSystemTime()` and steps forward a whole 30s period at a time until landing on a code provably absent from the accepted ±1-step set at that frozen instant, eliminating both the step-boundary race and the ~1-in-10^6 collision flake. Timers are restored in a `finally` block.
